### DISCUSSION TITLE
How to keep original order of attributes when Mojo::DOM print content ?

### DISCUSSION THREAD

**Discussion by vicyang** [mojolicious/mojo#2096] [General]
Hi friends,

example:

```perl
use Mojo::DOM;
my $str = qq(<p name="test" b="2" a="1"></p>);
print Mojo::DOM->new( $str )->to_string;
```

the result:
```html
<p a="1" b="2" name="test"></p>
```
The attributes are sorted alphabetically, how to keep the original order?

**Comment by Akron** [comment]
Hi! When parsing an XML document using `Mojo::DOM` XML attributes are stored in a hash map, so they have no particular order.This is in line with the XML specification. When the parsed DOM tree is rendered again (`to_string`), the attributes are sorted alphabetically for simplicity (testing, comparing etc.) only - there is no difference semantically.
So - in short: the order is not kept and therefore can not be replicated. Because both documents are semantically equivalent, your system should probably not rely on the attributes order at all, but for testing you can rely on the alphabetical order of rendered `Mojo::DOM` documents.

**Reply by vicyang** [reply]
Hi Akron, thank you for your reply
My work is modify the XML code for customer,
We often need to manually verify (compare) the details of the modified XML code, so the order of attributes is quite important for us.

My current Solution is to replace the order of certain specific attributes with regexp before export xml
```perl
$final_xml =~s/(<[a-zA-Z]+)( [^>]+)( name="[^"]+")/$1$3$2/g;
```

**Reply by Akron** [reply]
To verify/compare XML documents it is recommended to either use an XML compliant diff implementation or first turn both documents into something like [canonical XML](https://en.wikipedia.org/wiki/Canonical_XML). There are more possible but semantically meaningless differences between XML instances other than order of attributes (like whitespaces, newlines etc.). Using only Mojo, you could compare the rendering of both parsed `Mojo::DOM` documents (though I am not sure this will lead to fully canonical XML renderings, as `Mojo::DOM` is not only an XML but also an HTML5 parsing/rendering library).

But - of course - if changing the order of particular attributes prior to comparing already fits your needs, it may be sufficient!

**Comment by kraih** [comment]
Preserving attribute order would require big changes to the internal data structures in `Mojo::DOM` with API breakage, or the use of ordered hashes, which present a significant performance loss. Afraid i don't see it happen. The [JavaScript port of Mojo::DOM](https://www.npmjs.com/package/@mojojs/dom) does actually preserve order of attributes, as a side effect of how the language runtime works.
